src/storage/local.rs

```diff
@@ -53,10 +54,22 @@ impl Backend for LocalBackend {
         })
     }
     
-    fn upload_artifact_dir(&self, _name: &str, _version: u32, _env: Option<&str>) -> LalResult<()> {
+    fn upload_artifact_dir(&self, name: &str, version: u32, env: Option<&str>) -> LalResult<()> {
         // this fn basically assumes all the sanity checks have been performed
         // files must exist and lockfile must be sensible
-        let _artdir = Path::new("./ARTIFACT");
+        let artifactdir = Path::new("./ARTIFACT");
+        let tarball = artifactdir.join(format!("{}.tar.gz", name));
+        let lockfile = artifactdir.join("lockfile.json");
+
+        // prefix with environment
+        let prefix = env.map(|s| format!("env/{}/", s)).unwrap_or_else(|| "global/".into());
+        let tar_dir = format!("{}{}/{}/", prefix, name, version);
+        let tar_path = format!("{}{}/{}/{}.tar.gz", prefix, name, version, name);
+        let lock_path = format!("{}{}/{}/lockfile.json", prefix, name, version);
+
+        ensure_dir_exists_fresh(config_dir().join(tar_dir).to_str().unwrap())?;
```

Comment:
`.to_str().unwrap()` is probably unnecessary. Paths have a `.display()` method. But see first suggestion.